### PR TITLE
Fare rule entities -- internal representation

### PR DESCRIPTION
**Pull request by lionel-nj**
**Summary:**

This PR relates to processing of entities from fare_rules.txt. The goal is to use the required use cases implemented via PR #44 that allow to go from a ParsedEntity representing a row from fare_rules.txt to a specific concrete internal representation.

See #44 

**Expected behavior:** 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [ ] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/FareRule.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    @NotNull
+    final String fareId;
+
+    private final String routeId;
+    private final String originId;
+    private final String destinationId;
+    private final String containsId;
+
+    public FareRule(@NotNull final String fareId,
```
Comment (fabrice-v) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/FareRule.java]: Make private to enforce Builder usage

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/FareRule.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return destinationId;
+    }
+
+    public String getContainsId() {
+        return containsId;
+    }
+
+    public static class FareRuleBuilder {
+
+        @SuppressWarnings("CanBeFinal")
```
Comment (fabrice-v) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/FareRule.java]: Remove and provide setter for all fields

File: adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java [hunk 3]
```diff
@@ -196,4 +206,46 @@ public Level addLevel(final Level newLevel) throws IllegalArgumentException {
...
+     * @param originId      3rd part of the composite key: identifies an origin zone
+     * @param destinationId 4th part of the composite key: identifies a destination zone
+     * @param containsId    5th part ot the composite key: identifies the zones that a rider will enter while using a
+     *                      given fare class
+     * @return the FareRule representing a row from fare_rules.txt related to the id provided as parameter
+     */
+    @Override
+    public FareRule getFareRule(final String fareId, final String routeId, final String originId,
+                                final String destinationId, final String containsId) {
+        return fareRuleCollection.get(fareId + "; " + routeId + "; " + originId + "; " + destinationId + "; " +
```
[the remark this answers is not in the document]
Comment (lionel-nj) [reply adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java:248]: `;` is actually in the domain of definition of these fields. Good point. Thanks for the remark. Modifying this.

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/gtfs/FareRule.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            }
+        }
+    }
+
+    /**
+     * Returns the key corresponding to this {@link FareRule}
+     *
+     * @return the key corresponding to this {@link FareRule}
+     */
+    public String getFareRuleMappingKey() {
```
[the remark this answers is not in the document]
Comment (fabrice-v) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/gtfs/FareRule.java]: I see the changes, having both a static method and an instance one with the same name doesn't fit very well I think. I tried getting rid of the instance one by simply adding a public static String getFareRuleMappingKey(final FareRule buildKeyFrom) but couldn't make the test to work. I guess we can start and open an issue #285

**Pull request merged**